src/Functions/array/arrayMannWhitneyUTest.h

```diff
@@ -0,0 +1,201 @@
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+enum class Alternative {
+    TwoSided,
+    Greater,
+    Less
+};
+
+
+/** Calculates Mann-Whitney U nonparametric statistical test of the null hypothesis that it is
+  * equally likely that a randomly selected value from one population will be less than or greater
+  * than a randomly selected value from a second population. This test can be used to investigate
+  * whether two independent samples were selected from populations having the same distribution
+  *
+  * The function takes two arrays of samples
+  * These values are considered as the control and test samples of some observation
+  *
+  * https://en.wikipedia.org/wiki/Mann–Whitney_U_test
+  *
+  * Function returns p-value assuming an asymptotic normal distribution.
+  */
+
+
+template <typename Name, Alternative alternative>
+class ArrayMannWhitneyUTestImpl
+{
+private:
+    /// Initially allocate a piece of memory for 512 elements. NOTE: This is just a guess.
+    static constexpr size_t INITIAL_SIZE_DEGREE = 9;
+
+public:
+    static constexpr auto name = Name::name;
+    using ResultType = Float64;
+
+    static String getName()
+    {
+        return name;
+    }
+
+    static DataTypePtr getReturnType(const DataTypePtr &v1, const DataTypePtr &v2)
+    {
+        if (!(isNumber(v1) && isNumber(v2)))
+            throw Exception(std::string(getName()) + " label must have numeric type.", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        return std::make_shared<DataTypeNumber<ResultType>>();
+    }
+
+
+    template <class ForwardIt, class T>
+    static void iota(
+        ForwardIt first,
+        ForwardIt last,
+        T value)
+    {
+        while(first != last)
+        {
+            *first++ = value;
+            ++value;
+        }
+    }
+
+
+    template <typename T>
+    static PODArrayWithStackMemory<size_t, 1ULL << INITIAL_SIZE_DEGREE> sortIndexes(
+        const PODArrayWithStackMemory<T, 1ULL << INITIAL_SIZE_DEGREE> &v)
+    {
+        PODArrayWithStackMemory<size_t, 1ULL << INITIAL_SIZE_DEGREE> idx(v.size());
+        iota(idx.begin(), idx.end(), 0);
+
+        std::sort(idx.begin(), idx.end(),
+                  [&v](size_t i1, size_t i2) { return v[i1] < v[i2]; });
+
+        return idx;
+    }
+
+
+    template <typename T>
+    static PODArrayWithStackMemory<Float64, 1ULL << INITIAL_SIZE_DEGREE> rankData(
+        const PODArrayWithStackMemory<T, 1ULL << INITIAL_SIZE_DEGREE> &v)
+    {
+        PODArrayWithStackMemory<size_t, 1ULL << INITIAL_SIZE_DEGREE> idx = sortIndexes(v);
+        PODArrayWithStackMemory<Float64, 1ULL << INITIAL_SIZE_DEGREE> ranks(v.size());
+
+        for (size_t i = 0; i < v.size(); ++i)
+            ranks[idx[i]] = i + 1;
+
+        struct RankPair
+        {
+            size_t first;
+            Int64 second;
+        };
+
+        ClearableHashMap<T, RankPair, DefaultHash<T>, HashTableGrower<INITIAL_SIZE_DEGREE>,
```

Comment:
The size for allocator is incorrect. Please rebase to master, I added a helper for this -- `ClearableHashMapWithStackMemory`, it will set the right allocator size.

Reply:
(see this commit https://github.com/ClickHouse/ClickHouse/commit/0db71f52a37f73137ff0d1ae76f7a6125792eddf)